### ISSUE TITLE
Regression of the `typenum` crate on nightly

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#32637]
The [typenum crate](https://github.com/paholg/typenum) (as of https://github.com/paholg/typenum/commit/10c15e3fc3afd43e9b3d9070145f24b3fc7caec9 at least) successfully compiles on stable/beta but fails to compile no nightly with the error message:

```
error: overflow evaluating the requirement `uint::UInt<_, _> : private::PrivateCmp<uint::UInt<_, _>, Equal>` [E0275]
src/lib.rs:1:1: 1:1 note: consider adding a `#![recursion_limit="128"]` attribute to your crate
error: aborting due to previous error
```

It this expected? Perhaps this was bumping "close to the recursion limit" beforehand? I haven't diagnosed this much, but at least the reproduction is easy!

cc @aturon (maybe specialization related)
cc @rust-lang/compiler

**Comment by pnkfelix** [comment]
triage: P-high

**Comment by michaelwoerister** [comment]
Might be fixed by https://github.com/rust-lang/rust/pull/32579 once it lands.

**Comment by llogiq** [comment]
Nöte that [typenum PR #60](https://github.com/paholg/typenum/pull/60) fixes the problem from the typenum side.

**Comment by nikomatsakis** [comment]
@llogiq do you believe this to be a bug?

**Comment by llogiq** [comment]
Well, at best it's an unfortunate chain of events that unusual code may trigger, but that can be worked around. At worst, yes, it's a bug that can be worked around. If we deem it the latter, it's probably not too high on my wishlist.

**Comment by pnkfelix** [comment]
This particular instance of this problem seems to have resolved itself on `master`:

```
% git remote -v
origin  git@github.com:paholg/typenum.git (fetch)
origin  git@github.com:paholg/typenum.git (push)
% git status
HEAD detached at 10c15e3
nothing to commit, working directory clean
% multirust default nightly-2016-04-07
multirust: using existing install for 'nightly-2016-04-07'
multirust: default toolchain set to 'nightly-2016-04-07'
% rustc --version
rustc 1.9.0-nightly (bf5da36f1 2016-04-06)
% OUT_DIR=/home/fklock/Dev/Rust/typenum/target/debug RUSTC=~/Dev/Mozilla/rust.gi\
t/objdir-dbgopt/x86_64-unknown-linux-gnu/stage1/bin/rustc rustc src/lib.rs --cra\
te-name typenum --crate-type lib -g --out-dir /home/fklock/Dev/Rust/typenum/targ\
et/debug --emit=dep-info,link -L dependency=/home/fklock/Dev/Rust/typenum/target\
/debug -L dependency=/home/fklock/Dev/Rust/typenum/target/debug/deps
error: overflow evaluating the requirement `uint::UInt<_, _> : private::PrivateC\
mp<uint::UInt<_, _>, Equal>` [E0275]
src/lib.rs:1:1: 1:1 note: consider adding a `#![recursion_limit="128"]` attribut\
e to your crate
error: aborting due to previous error
% multirust default nightly-2016-04-08
multirust: using existing install for 'nightly-2016-04-08'
multirust: default toolchain set to 'nightly-2016-04-08'
% rustc --version
rustc 1.9.0-nightly (7979dd608 2016-04-07)
% OUT_DIR=/home/fklock/Dev/Rust/typenum/target/debug RUSTC=~/Dev/Mozilla/rust.gi\
t/objdir-dbgopt/x86_64-unknown-linux-gnu/stage1/bin/rustc rustc src/lib.rs --cra\
te-name typenum --crate-type lib -g --out-dir /home/fklock/Dev/Rust/typenum/targ\
et/debug --emit=dep-info,link -L dependency=/home/fklock/Dev/Rust/typenum/target\
/debug -L dependency=/home/fklock/Dev/Rust/typenum/target/debug/deps
%
```

Looking over the git log for rust in range `bf5da36f1..7979dd608`, my guess is that it was probably fixed by #32748 

Closing as fixed.
